### PR TITLE
Refactor: Extract common exporter logic into TrackingExporter base class

### LINKED ISSUE #11060: Langfuse exporter: No trace data found for span
This issue was created from Discord post 1448180612375777441:

[Open in Discord](https://discord.com/channels/1309558646228779139/1448180612375777441)

We are seeing a lot of warning when using Langfuse. 

Example warning: 
"WARN [2025-12-09 21:04:52.859 -0800] (Mastra): Langfuse exporter: No trace data found for span
    traceId: "ca1ea07b4fc6b259b397bc95957814ed"
    spanId: "53f0c3bf1a8574e1"
    spanName: "step: 0"
    spanType: "model_step"
    isRootSpan: false
    method: "handleSpanStarted""

I found a github issue that is supposed to fix this: https://github.com/mastra-ai/mastra/issues/8041 but I have all latest packages:

    "@mastra/core": "^0.24.7",
    "@mastra/evals": "^0.14.4",
    "@mastra/langfuse": "^0.2.3",
    "@mastra/libsql": "^0.16.4",
    "@mastra/memory": "^0.15.13",
    "@mastra/pg": "^0.17.10"

### PR DESCRIPTION
**Pull request by epinzur**
## Description

This PR consolidates the common span/event tracking logic from all vendor-specific observability exporters (Braintrust, Langfuse, LangSmith, PostHog) into a new TrackingExporter base class.

### Key Changes

New TrackingExporter base class (observability/mastra/src/exporters/tracking.ts)
  - Generic base class with type parameters for root, span, event, and metadata types
  - Handles trace data caching in a local memory map with TraceData class
  - Implements early data queue for out-of-order span arrival (when children arrive before parents)
  - Delayed cleanup scheduling to handle late-arriving data
  - Configurable memory management with soft/hard caps on trace counts
  - Abstract methods for vendor implementations: _buildRoot, _buildSpan, _buildEvent, _updateSpan, _endSpan, _abortSpan

New test utilities package (observability/_test-utils/)
  - trace-generator.ts - Generate traces with configurable depth/breadth, shuffle order
  - test-exporter.ts - TestTrackingExporter for isolated unit testing
  - test-scenarios.ts - Shared test scenarios for early data handling

Refactored exporters - All now extend TrackingExporter:
  - Removed ~300-400 lines of duplicated code from each exporter
  - Consistent behavior for edge cases (out-of-order spans, cleanup, etc.)
  - Added tracing.early-data.test.ts for each exporter

## Related Issue(s)

Fixes: #11060, #11694, #11543, #11661

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [X] Code refactoring
- [ ] Performance improvement
- [ ] Test update

## Checklist

- [ ] I have made corresponding changes to the documentation (if applicable)
- [X] I have added tests that prove my fix is effective or that my feature works

## Summary by CodeRabbit

* **New Features**
  * Introduced a generic TrackingExporter and migrated Braintrust, Langfuse, LangSmith, PostHog, and related exporters for robust out-of-order, late-trace, and cleanup handling.
  * Added observability test-utils: synthetic trace generator, test exporter harness, and shared early-data scenarios.
  * Added a usage-metrics formatter and a typed SpanErrorInfo.

* **Tests**
  * Large suite additions and refactors for early/late data, deep hierarchies, orphaned spans, and exporter behaviors; test-only helpers surface trace state.

* **Chores**
  * Added observability test-utils as a dev dependency across multiple packages.

* **Bug Fixes**
  * Safer environment restoration in tests and standardized disable handling during exporter init.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Review by abhiaiyer91 — APPROVED** [review]

File: observability/mastra/src/exporters/tracking.ts [hunk 1]
```diff
@@ -0,0 +1,1311 @@
...
+        // Note: Root span still continues to handleSpanStart below to track
+        // the span as active and call _buildSpan for vendor-specific handling
+      } else {
+        this.logger.debug(`${this.name}: Root does not exist, adding span to waiting queue.`, {
+          traceId: exportedSpan.traceId,
+          spanId: exportedSpan.id,
+        });
+        traceData.addToWaitingQueue({ event, waitingFor: 'root' });
+        return;
+      }
```
[the remark this answers is not in the document]
Comment (epinzur) [reply observability/mastra/src/exporters/tracking.ts:1089]: this is by design

**Review by wardpeet — APPROVED** [review]
